### PR TITLE
doc: add pre/postconditions and invariants to module keytest

### PR DESCRIPTION
**Pull request by lawli3t**
## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

### CODE REVIEW

File: src/libs/elektra/keyset.c [hunk 1]
```diff
@@ -576,6 +576,11 @@ Key *k2 = keyNew("user:/b", KEY_END);
  * Do not strcmp the keyName() yourself, because
  * the result differs from simple ascii comparison.
  *
+ * @pre The Keys @p k1 and @p k2 have been properly initialized via keyNew() or are NULL
+ * @invariant All parts of the Keys remain unchanged
```
Comment (markus2330) [inline src/libs/elektra/keyset.c:580]: What about ref counters?

File: src/libs/elektra/keyset.c [hunk 2]
```diff
@@ -576,6 +576,11 @@ Key *k2 = keyNew("user:/b", KEY_END);
...
+ * @post All parts of the Keys are unchanged
```
Comment (markus2330) [inline src/libs/elektra/keyset.c:581]: If it is already an invariant, the post condition is not needed anymore.

File: src/libs/elektra/keyset.c [hunk 3]
```diff
@@ -576,6 +576,11 @@ Key *k2 = keyNew("user:/b", KEY_END);
...
+ * @post If the result is 0, @p k1 and @p k2 cannot be used in the same KeySet
```
Comment (markus2330) [inline src/libs/elektra/keyset.c:582]: I don't understand the statement. Do you mean the return value when you say result? What means "cannot be used"?

To be more clear we should speak about the key names as the identity does not matter.

File: src/libs/elektra/keytest.c [hunk 4]
```diff
@@ -33,6 +33,11 @@
  * @ingroup key
  * @brief Methods to do various tests on Keys
  *
+ * With exception of the parameters of keyCmp(), the following contract holds for all parameters of type Key:
+ * @pre The Key has been properly initialized via keyNew()
+ * @invariant All parts of the Key remain unchanged
+ * @post All parts of the Key are unchanged
```
Comment (markus2330) [inline src/libs/elektra/keytest.c:39]: Implied by invariant

```suggestion
```

**Review by markus2330 — COMMENTED** [review]
Thank you, great to see this!

File: src/libs/elektra/keyset.c (hunk 2 above)
Comment (mpranj) [inline src/libs/elektra/keyset.c:581]: I agree, this is redundant.
```suggestion
```

File: src/libs/elektra/keyset.c (hunk 1 above)
Comment (mpranj) [reply src/libs/elektra/keyset.c:580]: Doesn't look like `keyCmp ()` changes the refcounters.

File: src/libs/elektra/keyset.c (hunk 3 above)
Comment (mpranj) [reply src/libs/elektra/keyset.c:581]: Can be improved in a new PR.

**Pull request merged**